_posts/2018-01-01-futures-0.3.0-alpha.3.md

```diff
@@ -0,0 +1,73 @@
+
+`spawn!` lets you spawn futures with `Output = ()`:
+
+```rust
+#![feature(async_await, await_macro, futures_api)]
+#[macro_use] extern crate futures;
+
+spawn!(future).unwrap();
+```
+
+`spawn_with_handle!` lets you spawn futures with any `Output` type and also lets you await the output:
+
+```rust
+let join_handle = spawn_with_handle!(future).unwrap();
+let output = await!(join_handle);
+```
+
+`spawn!` is slightly cheaper than `spawn_with_handle!`.
+
+Should spawning fail, a `SpawnError` error is returned which contains information about why it failed.
+
+So what futures can be spawned? There are some requirements: The future needs to be `Send` and `'static` and the output needs to be `Send`. Here's why:
+- The `'static` lifetime means that the future cannot contain any non-static (exterior) references. This is required so that the spawned task can be run independently to the task that spawns it.
+- The `Send` bounds mean that the future and its output can be safely sent to other operating system threads. This is required so that multithreaded executors can run the spawned task on any operating thread they choose.
+
+Note: You can execute non-`Send`, non-`'static` futures concurrently via the `join!` macro.
+
+For use outside of async functions, there's also something new: The `ExecutorExt` trait. It provides the methods `spawn` and `spawn_with_handle`:
+
+```rust
+use futures::task::ExecutorExt;
+
+// spawn()
+executor.spawn(future).unwrap();
+
+// spawn_with_handle()
+let join_handle = executor.spawn_with_handle(future).unwrap();
+```
+
+Inside `poll`, you can use these methods by accessing the executor via `Context::executor`:
+
+```rust
+fn poll(self: PinMut<Self>, cx: &mut task::Context) -> Poll<T> {
+    cx.executor().spawn(future);
+}
+```
+
+## The plan: `spawn` with `dyn Future`
+
+We plan to eventually replace `ExecutorExt::spawn` with `Executor::spawn`. This method will live directly in the `Executor` trait defined in libcore. It's signature will look like this:
+
+```rust
+fn spawn(
+    &mut self,
+    future: dyn Future<Output = ()> + Send + 'static,
+) -> Result<(), SpawnError>
+```
+
+This method can't be implemented just yet due to compiler limitations:
+- Arbitrary self types are currently not object-safe: `Future::poll` uses `PinMut` as arbitrary self type. This unfortunately makes it impossible to create trait objects for `Future`. Therefore, `Box<dyn Future>` and `&dyn Future` are *currently* not supported.
+- Unsized rvalues are not yet available: Trait objects (`dyn Trait`) are `!Sized` because the size of a trait object depends on the size of the type inside it and can therefore only be known at runtime. Currently we often use `Box<dyn Trait>` which stores the trait object on the heap to circumvent that. The unsized rvalues feature will make it possible to store `!Sized` types on the stack.
```

Comment:
I don't feel that "the type inside it" is really correct in "the size of a trait object depends on the size of the type inside it", trait objects aren't really wrappers. Maybe something like "the size of a trait object depends on the size of the concrete type behind it"

Reply:
They're not? Aren't `dyn Trait` basically the following?
```rust
pub struct TraitObject {
    vtable: *mut (),
    data: Type,
}
```
If yes, then my formulation should work. But, yours works as well, so I'm going to change it.

Reply:
Except in cases other than unsized rvalues the vtable is split from the data out into the pointer type, e.g. `Box<dyn Trait>` is similar to

```
struct Box<dyn Trait> {
    vtable: *mut (),
    data: *mut Type,
}
```

so a "trait object" isn't a real construction in the type system, rather more like an idea that is implemented differently depending on the wrapper type. Once there are unsized rvalues I assume they'll have a construction similar to your example, but that's specifically the unsized rvalue case, not the "trait object" itself. One alternative name I've seen for unsized rvalues is "bare trait object", but that seems to have been co-opted by the lint for renaming `Trait -> dyn Trait`, so I'm not sure if there's a nice way to refer specifically to "unboxed trait objects" now.